src/Common/src/System/Security/Cryptography/RsaPaddingProcessor.cs

```diff
@@ -0,0 +1,557 @@
+            }
+
+            // 7. dbMask = MGF(H, emLen - hLen - 1)
+            byte[] dbMaskRented = ArrayPool<byte>.Shared.Rent(maskedDb.Length);
+            Span<byte> dbMask = new Span<byte>(dbMaskRented, 0, maskedDb.Length);
+
+            try
+            {
+                using (IncrementalHash hasher = IncrementalHash.CreateHash(_hashAlgorithmName))
+                {
+                    Mgf1(hasher, h, dbMask);
+
+                    // 8. DB = maskedDB XOR dbMask
+                    for (int i = 0; i < dbMask.Length; i++)
+                    {
+                        dbMask[i] ^= maskedDb[i];
+                    }
+
+                    // 9. Set the unused bits of DB to 0
+                    dbMask[0] &= usedBitsMask;
+
+                    // 10 ("a"): If the emLen - hLen - sLen - 2 leftmost bytes are not 0,
+                    // output "inconsistent" and stop.
+                    //
+                    // Since signature verification is a public key operation there's no need to
+                    // use fixed time equality checking here.
+                    for (int i = emLen - _hLen - sLen - 2 - 1; i >= 0; --i)
+                    {
+                        if (dbMask[i] != 0)
+                        {
+                            return false;
+                        }
+                    }
+
+                    // 10 ("b") If the octet at position emLen - hLen - sLen - 1 (under a 1-indexed scheme)
+                    // is not 0x01, output "inconsistent" and stop.
+                    if (dbMask[emLen - _hLen - sLen - 2] != 0x01)
+                    {
+                        return false;
+                    }
+
+                    // 11. Let salt be the last sLen octets of DB.
+                    ReadOnlySpan<byte> salt = dbMask.Slice(dbMask.Length - sLen);
+
+                    // 12/13. Let H' = Hash(eight zeros || mHash || salt)
+                    hasher.AppendData(s_eightZeros);
+                    hasher.AppendData(mHash);
+                    hasher.AppendData(salt);
+
+                    Span<byte> hPrime = stackalloc byte[_hLen];
+
+                    if (!hasher.TryGetHashAndReset(hPrime, out int hLen2) || hLen2 != _hLen)
+                    {
+                        Debug.Fail("TryGetHashAndReset failed with exact-size destination");
+                        throw new CryptographicException();
+                    }
+
+                    // 14. If H = H' output "consistent". Otherwise, output "inconsistent"
+                    //
+                    // Since this is a public key operation, no need to provide fixed time
+                    // checking.
+                    return h.SequenceEqual(hPrime);
+                }
+            }
+            finally
+            {
+                dbMask.Clear();
+                ArrayPool<byte>.Shared.Return(dbMaskRented);
+            }
+        }
+
+        // https://tools.ietf.org/html/rfc3447#appendix-B.2.1
```

Comment:
Yeah, this one is sort of weird to write in the order they wrote it in the RFC.

> 1. If maskLen > 2^32 hLen, output "mask too long" and stop.

We know that's not the case, because maskLen was specified as `mask.Length`

> 3. For counter from 0 to \ceil (maskLen / hLen) - 1, do the
      following:

It wants us to hash the minimum number of times we need to to produce maskLen bytes.

> 
      a. Convert counter to an octet string C of length 4 octets (see
         Section 4.1):
>
>            C = I2OSP (counter, 4) .

That means "make a 4 byte big endian value for `counter`".  Aka `BinaryPrimitives.WriteInt32BigEndian(bigEndianCount, count);`

>      b. Concatenate the hash of the seed mgfSeed and C to the octet
         string T:
>
>            T = T || Hash(mgfSeed || C) .

Let's scope it in to `Hash(mgfSeed || C)`:
```C#
hasher.AppendData(mgfSeed);
BinaryPrimitives.WriteInt32BigEndian(bigEndianCount, count);
hasher.AppendData(bigEndianCount);
```

Let's skip ahead:

>    4. Output the leading maskLen octets of T as the octet string mask.

Okay, so rather than build up a big T we can just write this out to the destination because we don't use the collective T value.

The only problem we have is when the last hash is bigger than it needs to be, so we need to write it to a temporary location and then only copy over the remaining data.

Now that we've seen how we got to the end, we look back at

>   2. Let T be the empty octet string.

And conclude that it was required for the formula, but not the implementation.